src/main/flight/imu.c

```diff
@@ -279,20 +243,19 @@ static void imuMahonyAHRSupdate(float dt, float gx, float gy, float gz,
 
         // For magnetometer correction we make an assumption that magnetic field is perpendicular to gravity (ignore Z-component in EF).
         // This way magnetic field will only affect heading and wont mess roll/pitch angles
-
         // (hx; hy; 0) - measured mag field vector in EF (assuming Z-component is zero)
         // (bx; 0; 0) - reference mag field vector heading due North in EF (assuming Z-component is zero)
-        const float hx = rMat[0][0] * mx + rMat[0][1] * my + rMat[0][2] * mz;
-        const float hy = rMat[1][0] * mx + rMat[1][1] * my + rMat[1][2] * mz;
+        const float hx = (1.0f - 2.0f * (qpAttitude.yy + qpAttitude.zz)) * mx + (2.0f * (qpAttitude.xy - qpAttitude.wz)) * my + (2.0f * (qpAttitude.xz + qpAttitude.wy)) * mz;
+        const float hy = (2.0f * (qpAttitude.xy + qpAttitude.wz)) * mx + (1.0f - 2.0f * (qpAttitude.xx + qpAttitude.zz)) * my + (2.0f * (qpAttitude.yz - qpAttitude.wx)) * mz;
         const float bx = sqrtf(hx * hx + hy * hy);
 
         // magnetometer error is cross product between estimated magnetic north and measured magnetic north (calculated in EF)
         const float ez_ef = -(hy * bx);
 
         // Rotate mag error vector back to BF and accumulate
-        ex += rMat[2][0] * ez_ef;
```

Comment:
Only rotation in one axis is important. With quaternion / generic code full cross product should be computed and than x,y zeroed, then used for rotation ...